### PR TITLE
Add ragged paged attention

### PR DESCRIPTION
**Pull request by vanbasten23**
Test plan:
```
LIBTPU_INIT_ARGS=--xla_tpu_scoped_vmem_limit_kib=65536  python /workspaces/persist/pytorch/xla/test/test_ragged_paged_attention_kernel.py 2>&1 | tee out.txt
```

cc: @miladm

### CODE REVIEW

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py [hunk 1]
```diff
@@ -0,0 +1,976 @@
...
+  num_logical_q_blks = pl.num_programs(1)
+  num_q_heads_per_kv_head, num_queries_per_compute_block, head_dim = q_ref.shape
+  num_kv_heads, total_num_pages, page_size, head_dim = k_pages_hbm_ref.shape
+  kv_blk_size = page_size * num_kv_pages_per_compute_block
+  
+  seq_ids, physical_q_tile_ids = group_metadata_ref
+  cur_seq_idx = seq_ids[logical_q_blk_idx]
+  effective_kv_len_cur_seq = effective_kv_lens_ref[cur_seq_idx]
+  should_run = (kv_blk_idx * kv_blk_size < effective_kv_len_cur_seq)
+  pl.debug_print("xw32 paged_flash_attention_kernel begins kv_head_idx={}, logical_q_blk_idx={}, kv_blk_idx={}, cur_seq_idx={}, effective_kv_len_cur_seq={}", kv_head_idx, logical_q_blk_idx, kv_blk_idx, cur_seq_idx, effective_kv_len_cur_seq)  # pl.debug_print can only print JAX type. So cannot print tuple such as q.shape.
```
Comment (bythew3i) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py]: We should remove all the `pl.debug_print` in prod. It will emit instructions in TPU and waste memory for storing debug data and message.

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py [hunk 2]
```diff
@@ -0,0 +1,976 @@
...
+  # Need to store these l_next and m_next which will relay to the output.
+  # But only update the part that belongs to the current sequence we are working on.
+  lm_mask = _get_store_mask(grid_id=logical_q_blk_idx,
+                            group_offsets=effective_cu_q_lens_ref,
+                            group_ids=seq_ids,
+                            m_tile_ids=physical_q_tile_ids,
+                            tm=num_queries_per_compute_block,
+                            tn=MIN_BLOCK_SIZE,
+                            )
+  # Should I use jax.lax.select or jnp.where? What's the difference? eg: jnp.where(lm_mask, l_next, 0), jnp.where(lm_mask, m_next, 0)
```
Comment (bythew3i) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py]: `jnp.where` wraps `jax.lax.select`

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py [hunk 3]
```diff
@@ -0,0 +1,976 @@
...
+              "arbitrary",
+          )),
+      out_shape=out_shape,
+  )
+  # TODO: need to slice the page_indices later to avoid the SMEM OOM.
+  page_indices_1d = page_indices.reshape(-1)
+  buffer_index = jnp.zeros((1,), jnp.int32)
+  step = jnp.zeros((1,), jnp.int32)
+
+  # debug compile begins
```
Comment (bythew3i) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py:900]: Remove?

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py [hunk 4]
```diff
@@ -0,0 +1,976 @@
...
+  num_logical_q_blks = pl.num_programs(1)  # grid=(num_kv_heads, num_logical_q_tiles, num_kv_blks)
+  next_logical_q_blk_idx = jnp.where(logical_q_blk_idx == num_logical_q_blks - 1,
+                                     logical_q_blk_idx,
+                                     logical_q_blk_idx+1)
+  is_last_logical_q_blk = (logical_q_blk_idx == num_logical_q_blks-1)
+  physical_q_blk_will_change = (physical_q_tile_ids[logical_q_blk_idx] != physical_q_tile_ids[next_logical_q_blk_idx])
+  last_time_seeing_cur_physical_q_blk = jnp.logical_or(is_last_logical_q_blk, physical_q_blk_will_change)
+  should_store_to_hbm = jnp.logical_and(is_last_kv_blk_idx, last_time_seeing_cur_physical_q_blk)
+  @pl.when(should_store_to_hbm)
+  def store_to_hbm():  # pylint: disable=unused-variable
```
Comment (bythew3i) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py:509]: This function actually only stores to VMEM and all the ref used are actually VMEM ref. We rely on pipeline emitter to send vmem block back to HBM. And we can't store vregs directly to HBM in kernel. So I think original store_to_output makes more sense here.

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py [hunk 5]
```diff
@@ -0,0 +1,976 @@
...
+  ]
+
+  kernel = pl.pallas_call(
+      functools.partial(
+          paged_flash_attention_kernel,
+          pages_per_sequence=pages_per_sequence,
+          num_tokens=num_tokens,
+          num_seqs=num_seqs,  # it they changes, need to recompile.
+          num_kv_pages_per_compute_block=num_kv_pages_per_compute_block,
+          mask_value=mask_value,
```
Comment (bythew3i) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py]: These are all calculated from either static shape or static_argnames. 

The overhead could be very large if some are too dynamic (num_seqs, num_tokens) because we need to recompile many times. Maybe to add a note or TODO here.

File: test/test_ragged_paged_attention_kernel.py [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+    num_seqs = 3
+    metadata, num_logical_q_tiles  = make_group_metadata(
+        cu_q_lens=cu_q_lens,
+        m=num_q_tokens,
+        tm=num_queries_per_compute_block,
+        start_group=start_group,
+        num_seqs=num_seqs
+    )
+    seq_ids, physical_q_tile_ids = metadata
+    # print(f"xw32 metadata.physical_q_tile_ids: {metadata.physical_q_tile_ids}")
```
[the remark this answers is not in the document]
Comment (vanbasten23) [reply test/test_ragged_paged_attention_kernel.py]: hmm, I don't see it in my code? Is it from an old commit?

**Comment by bythew3i** [comment]
> Test plan:
> 
> ```
> LIBTPU_INIT_ARGS=--xla_tpu_scoped_vmem_limit_kib=65536  python /workspaces/persist/pytorch/xla/test/test_ragged_paged_attention_kernel.py 2>&1 | tee out.txt
> ```

How is 65536 calculated?

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py [hunk 7]
```diff
@@ -0,0 +1,976 @@
...
+    cu_q_lens: jnp.ndarray,
+    m: int,
+    tm: int,
+    start_group: jnp.ndarray,
+    num_seqs: int,
+):
+  """Create the metadata needed for grouped matmul computation.
+
+  Args:
+    group_sizes: A 1d, jnp.ndarray with shape [num_groups] and jnp.int32 dtype.
```
[the remark this answers is not in the document]
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py]: I don't see it in my code. Maybe it's from an old commit?

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py [hunk 8]
```diff
@@ -0,0 +1,976 @@
...
+  if kv_lens.shape[0] != num_tokens:
+    raise ValueError("kv_lens.shape[0] must be thet same as num_tokens. Got"
+                     f" {kv_lens.shape[0]} and {num_tokens}")
+  if page_indices.shape[0] != num_tokens:
+    raise ValueError("page_indices.shape[0] must be thet same as num_tokens. Got"
+                     f" {page_indices.shape[0]} and {num_tokens}")
+  if cu_q_lens.shape[0] != num_tokens + 1:
+    raise ValueError("cu_q_lens.shape[0] must be thet same as num_tokens + 1. Got"
+                     f" {cu_q_lens.shape[0]} and {num_tokens + 1}")
+  if num_seqs > num_tokens:
```
[the remark this answers is not in the document]
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py:287]: Let's say there are num_tokens=N, then
One extreme is there is one sequence and all the N tokens are in this sequence. (num_seqs=1, num_tokens=N)
The other extreme is each sequence has only 1 tokens. (num_seqs==num_tokens==N).
So there won't a case where num_seqs>num_tokens

[the remark this answers is not in the document]
Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py:287]: I see so the num_tokens here is total num_tokens!

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py [hunk 9]
```diff
@@ -0,0 +1,976 @@
...
+    raise ValueError(f"num_seqs must be less or equal to num_tokens. Got {num_seqs} and {num_tokens}")
+  # int16: will pack. need to explicit cast to int32. int64 is not supported in Pallas. for smem 1d case.
+  # 2d smem: int16 will be packed with an empty. So we didn't save any memory.
+  # scalar: use i32 (1, N). int16 for (1, N) will be padding. Need to use (2, N).
```
[the remark this answers is not in the document]
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py:293]: removed the comments.

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py [hunk 10]
```diff
@@ -0,0 +1,976 @@
...
+                              1.0 / l_next)  # [block_q, 128]
+  temp = acc_scratch_ref[q_head_idx_per_kv] * l_broadcast(l_corr * l_next_inv_safe)
+  acc_mask = _get_store_mask(grid_id=logical_q_blk_idx,
+                            group_offsets=effective_cu_q_lens_ref,
+                            group_ids=seq_ids,
+                            m_tile_ids=physical_q_tile_ids,
+                            tm=num_queries_per_compute_block,
+                            tn=head_dim,
+                            )
+  print(f"xw32 line486 {acc_mask.shape=}, {temp.shape=}, {acc_scratch_ref[q_head_idx_per_kv]=}")
```
[the remark this answers is not in the document]
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py]: I don't see it in my code. Maybe it's from an old commit?

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py [hunk 11]
```diff
@@ -0,0 +1,976 @@
...
+  # But only update the part that belongs to the current sequence we are working on.
+  lm_mask = _get_store_mask(grid_id=logical_q_blk_idx,
+                            group_offsets=effective_cu_q_lens_ref,
+                            group_ids=seq_ids,
+                            m_tile_ids=physical_q_tile_ids,
+                            tm=num_queries_per_compute_block,
+                            tn=MIN_BLOCK_SIZE,
+                            )
+  # Should I use jax.lax.select or jnp.where? What's the difference? eg: jnp.where(lm_mask, l_next, 0), jnp.where(lm_mask, m_next, 0)
+  # Can `lm_mask[...]` be `lm_mask`?
```
[the remark this answers is not in the document]
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py]: Do you know if `lm_mask[...]` is the same as `lm_mask`? Both seem to work.

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py [hunk 12]
```diff
@@ -0,0 +1,976 @@
...
+  l_scratch_ref[q_head_idx_per_kv] = jax.lax.select(lm_mask[...], l_next, l_scratch_ref[q_head_idx_per_kv])
+  m_scratch_ref[q_head_idx_per_kv] = jax.lax.select(lm_mask[...], m_next, m_scratch_ref[q_head_idx_per_kv])
```
[the remark this answers is not in the document]
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py]: Will do! Quick question, by "waste VPU", are you referring to the part when the mask is false, we assign `l_scratch_ref[q_head_idx_per_kv]` to `l_scratch_ref[q_head_idx_per_kv]`? If so, in this case what operations will VPU do?

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py]: Yes they are the same because lm_mask is vector (already loaded in vector mask registers).

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py:455]:
```
ref[...]= jax.lax.select(mask, data, ref[...])  -> pl.store(ref, ..., data, mask=mask)
```

The former generates 2 instructions (vselect + store) but the latter generates only 1 vstore.

[the remark this answers is not in the document]
Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py:455]: actually the former generates 2 instructions (vstore + vselect + vstore)

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py (hunk 1 above)
Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_kernel.py]: I don't see it in my code. Maybe it's from an old commit?

**Comment by vanbasten23** [comment]
> > Test plan:
> > ```
> > LIBTPU_INIT_ARGS=--xla_tpu_scoped_vmem_limit_kib=65536  python /workspaces/persist/pytorch/xla/test/test_ragged_paged_attention_kernel.py 2>&1 | tee out.txt
> > ```
> 
> How is 65536 calculated?

I found a ticket and someone uses it. I remember the number is the vmem limit on a TPU generation.

**Review by bythew3i — APPROVED** [review]

**Comment by vanbasten23** [comment]
`Build and test / CPU tests / test (benchmark_tests)` failure is irrelevant to this PR. (A PR https://github.com/pytorch/xla/pull/8668 without any changes also fails this)

**Review by miladm — APPROVED** [review]

File: test/tpu/run_tests.sh [hunk 13]
```diff
@@ -37,7 +37,8 @@ run_xla_hlo_debug python3 "$TEST_CDIR/scan/test_scan_debug.py"
 python3 "$TEST_CDIR/test_pallas.py" -v
 python3 "$TEST_CDIR/test_pallas_spmd.py"
 XLA_DISABLE_FUNCTIONALIZATION=1 python3 "$TEST_CDIR/test_pallas_spmd.py"
-python3 "$TEST_CDIR/test_tpu_paged_attention_kernel.py"
+python3 "$TEST_CDIR/test_multi_queries_paged_attention_kernel.py"
```
Comment (miladm) [inline test/tpu/run_tests.sh:40]: isn't this line renamed?
